packages/react-dom/src/__tests__/ReactCompositeComponent-test.js

```diff
@@ -1500,4 +1500,45 @@ describe('ReactCompositeComponent', () => {
     ReactTestUtils.renderIntoDocument(<Component />);
     expect(mockArgs.length).toEqual(0);
   });
+
+  it('should return a meaningful warning when constructor is returned', () => {
+    spyOn(console, 'error');
+    class RenderTextInvalidConstructor extends React.Component {
+      constructor(props) {
+        super(props);
+        return {something: false};
+      }
+
+      render() {
+        return <div />;
+      }
+    }
+
+    expect(function() {
+      ReactTestUtils.renderIntoDocument(<RenderTextInvalidConstructor />);
+    }).toThrow();
+
+    const error = console.error.calls.mostRecent().args[0];
```

Comment:
Let's also assert we only emit one warning. Like other tests do by asserting on count.